### PR TITLE
rewrite the world

### PR DESCRIPTION
**Pull request by nlf**
this module leaned too heavily on repeated parsing and regular expressions, seeing as we use it in hot paths everywhere it felt pretty important to make sure that it's as fast as is reasonable.

I modernized it to take advantage of features that were not available in node when it was originally written, I also modified the tests so that it's purely snapshot based and fixtures are programmatically generated urls based on a set of inputs. it does filter out duplicate urls so that we don't have to do too many repeated tests, and it also tests against a matrix of options for `noGitPlus` and `noCommittish`

I don't _love_ the snapshot organization, but it's good enough for now IMO and shows clearly what parses and what doesn't.

reviewers, brace yourselves.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -36,111 +36,107 @@ module.exports.fromUrl = function (giturl, opts) {
...
+
+// attempt to correct an scp style url so that it will parse with `new URL()`
+const correctUrl = (giturl) => {
+  const firstAt = giturl.indexOf('@')
+  let firstColon = giturl.indexOf(':')
+  let lastColon = giturl.lastIndexOf(':')
+
+  let corrected
+  if (lastColon > firstAt) {
+    // the last : comes after the first @ (or there is no @)
```
Comment (wraithgar) [inline index.js:98]:
```suggestion
    // if the last : comes after the first @ (or there is no @)
```

File: index.js [hunk 2]
```diff
@@ -36,111 +36,107 @@ module.exports.fromUrl = function (giturl, opts) {
...
-    hash: matched[4],
-    search: null,
-    query: null,
-    pathname: '/' + matched[3],
-    path: '/' + matched[3],
-    href: 'git+ssh://' + matched[1] + '@' + matched[2] +
-          '/' + matched[3] + (matched[4] || '')
+
+  if (firstColon === -1) {
+    // we have no : at all
```
Comment (wraithgar) [inline index.js:115]:
```suggestion
    // if we have no : at all
```

File: test/fixtures/failures.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  () => '',
+  () => null,
+  () => undefined
+]
+
+exports.params = [
+  {}
+]
+
+/* these need to be ported
```
Comment (wraithgar) [inline test/fixtures/failures.js:21]: Did this get done?

Comment (nlf) [reply test/fixtures/failures.js:21]: nope! good catch, forgot about those. I'll get them added in

**Comment by fregante** [comment]
I’d probably move the test changes to a separate PRs. Replacing 100% of the tests and code at once defeats the purpose of even having tests; Regressions won’t be caught this way.

**Comment by nlf** [comment]
> I’d probably move the test changes to a separate PRs. Replacing 100% of the tests and code at once defeats the purpose of even having tests; Regressions won’t be caught this way.

I agree, I started working on that already. there are more changes that will be coming here too, so I've converted this back to a draft PR while I get everything sorted out

**Pull request closed without merging**